### DISCUSSION TITLE
v2.5.0

### DISCUSSION THREAD

**Discussion by mmgoodnow** [cross-seed/cross-seed#112] [General]
- Merge pull request #111 from mmgoodnow/daemon-perf  9e72461
- Use absolute paths  6753042
- Finishing touches  5efbe79
- Finish preliminary implementation  f186f8a
- Convert some synchronous fs code to promises  e0d826b
- Remove dead function loadTorrentDir  f33ab69

https://github.com/mmgoodnow/cross-seed/compare/v2.4.1...v2.5.0

<hr /><em>This discussion was created from the release <a href='https://github.com/mmgoodnow/cross-seed/releases/tag/v2.5.0'>v2.5.0</a>.</em>

**Comment by mmgoodnow** [comment]
With 2.5.0, Daemon mode is now officially supported on qBittorrent! @Drazzilb08 @ajhorvath20 you might be interested in this. Please report any bugs you see!

Previously, daemon mode had the following restrictions:
- In order to find the torrent to be searched (internally, the "searchee torrent") when supplied a name, we had to read each torrent in your torrent dir one-by-one until we found one with the correct name. This could be very slow, and if you have 1000 torrents, on average it has to read 500 torrents, each of which can be hundreds of kilobytes, every time the webhook gets called.
- In order to generate the list of info hashes to exclude from search results, we avoided having to do the slow process of reading every torrent just to get its infohash, and instead we relied on the file naming scheme of the rTorrent session directory, which names torrents by infohash. However this meant that any directory where the torrents weren't named by infohash wouldn't work properly with existing torrent exclusion. 

We now index the torrents in your torrentDir in daemon mode. This means that the list of info hashes is generated beforehand and stored, and whenever new files are added to the torrent dir, cross-seed only needs to process the files that are new.

TL;DR: Daemon mode is now faster, does less disk reading, and is compatible with qBittorrent (and other future clients)!

**Reply by ajhorvath20** [reply]
Hi @mmgoodnow nice one, works perfectly fine 
```
today at 09:02:39 Configuration file already exists.
today at 09:02:39 info: Validating your configuration...
today at 09:02:39 info: Your configuration is valid!
today at 09:02:39 info: [server] Server is running on port 2468, ^C to stop.
today at 09:22:16 info: [server] Received name A.2021.1080p.WEB-DL.DD5.1.H.264-EVO.mkv
today at 09:22:19 info: Found A.2021.1080p.WEB-DL.DD5.1.H.264-EVO.mkv on tracker123 - exists
today at 09:22:19 info: [server] Found 1 torrents for A.2021.1080p.WEB-DL.DD5.1.H.264-EVO.mkv
```
Thanks

**Reply by Drazzilb08** [reply]
Hello @mmgoodnow Thank you for this update. :D
I've been out of the game for a bit, testing it right now. My collection has grown a bit. I've been running on `search` for quite some time w/o any issues. 
I'm messing with daemon mode right now. Quick question, is there anything I need to set up w/in qbit in order to enable this functionality?
Been looking over the docs and didn't see anything updated w/ respects to this.
```
today at 1:21:41 PM Container started
today at 1:21:42 PM Configuration file already exists.
today at 1:21:42 PM info: Validating your configuration...
today at 1:21:44 PM info: Your configuration is valid!
today at 1:21:44 PM info: [server] Server is running on port 2468, ^C to stop.
```
This is what I'm getting, just not sure if I have it all set up the intended way.

**Reply by afraazali** [reply]
That's looks right @Drazzilb08. From your qbit you'd need to run a script on completion that sends a request to the daemon. My qbit uses a VPN, so I couldn't do that with my current setup, so what I did was setup a custom script in radarr and sonarr that makes the API request to the daemon.

**Reply by Drazzilb08** [reply]
@afraazali Hmmm interesting, what script would be used in order to facilitate this functionality?
My qbit instance is on a VPN however, it is all local.

Side note, how does this functionality account for torrents propagating to other trackers, much like how I could tell search to give a couple days in between each scan to give that time for propagation and maximize cross seeding. This seems to be a more active searching method but only a one time deal.
Is this correct?

**Reply by afraazali** [reply]
This is a sample bash script I used for sonarr:

```
#!/bin/bash

release_hash=$sonarr_download_id
if [[ ${sonarr_eventtype^^} == "TEST" ]]; then
	release_hash="Test TV Release"
fi

curl -XPOST ${CROSS_SEED_HOST}/api/webhook \
  -H 'Content-Type: application/json' \
  --data '{"infoHash":"'"${release_hash}"'"}'

exit 0
```

This just makes a curl request to my cross seed daemon container using the release hash. I'm looking at the qbit documentation and it says you can use `%I` to substitute for the info hash. So maybe something like this would work:

```
curl -XPOST ${CROSS_SEED_HOST}/api/webhook \
  -H 'Content-Type: application/json' \
  --data '{"infoHash":"'"${1}"'"}'

exit 0
```

Of course fill in the cross seed host with whatever your domain/port is: eg `http://cross-seed:2468` and in qbit, the `Run external program on torrent completion` would be something like:

`/path/to/the/above/script %I`

I haven't tested it but that _should_ work

**Reply by Drazzilb08** [reply]
Okay, I'm understanding the concept of all of this. Thank you for the quick replies. I'll see how this could possibly be integrated into my set up.

**Reply by undaunt** [reply]
Just hopping in here to discuss - but I also have my qBittorrent running through a wireguard VPN (local, separate docker container) and I've found you can inject (haven't finished testing daemon mode, but in search mode at least) by passing through the wireguard container and ports qbittorrent is running through.

Here's my setup that works:

```
  wireguard:
    image: lscr.io/linuxserver/wireguard
    container_name: wireguard
    restart: always
    networks:
      - $TRAEFIK_NETWORK
    security_opt:
      - no-new-privileges:true
    cap_add:
      - NET_ADMIN
      - SYS_MODULE
    sysctls:
      - net.ipv4.conf.all.src_valid_mark=1
      - net.ipv6.conf.all.disable_ipv6=1
    ports:
      - 51820:51820
      - $WG_QB_PORT2:$WG_QB_PORT2 # qb-tv
    volumes:
      - $DOCKERDIR/wireguard:/config
      - /lib/modules:/lib/modules
    environment:
      - TZ
      - PUID
      - PGID
    labels:
      ## Disable watchtower automatic updates
      - com.centurylinklabs.watchtower.enable=false
      ## Enable Traefik
      - traefik.enable=true
      ## HTTP Routers
      - traefik.http.routers.qb-tv-rtr.entrypoints=https
      - traefik.http.routers.qb-tv-rtr.rule=Host(`tv.$DOMAINNAME`)
      ## Middlewares
      - traefik.http.routers.qb-tv-rtr.middlewares=chain-authelia@file
      ## HTTP Services
      - traefik.http.routers.qb-tv-rtr.service=qb-tv-svc
      - traefik.http.services.qb-tv-svc.loadbalancer.server.port=$WG_QB_PORT2
```

```
  qb-tv:
    image: lscr.io/linuxserver/qbittorrent:14.3.9
    container_name: qb-tv
    restart: unless-stopped
    network_mode: service:wireguard
    security_opt:
      - no-new-privileges:true
    depends_on:
      - wireguard
    volumes:
      - $DOCKERDIR/qb-tv:/config
      - $DLDIR:/downloads
    environment:
      - TZ
      - PUID
      - PGID
      - WEBUI_PORT=$WG_QB_PORT2 # Listening port
```

To accomplish port forwarding, inside qB config I enter the port the VPN provider has set up for forwarding, whereas all the ports above "$WG_QB_PORT2" is, in this case, 8082 for reverse proxying on Traefik.

Were you all able to finalize a successful qBittorrent injection script for daemon mode?

Thanks.

**Reply by undaunt** [reply]
Re: the above, I got the script to work. I think my bash script just needed a syntax or missing quotation. My qB is calling on completion: `/config/cross-seed.sh "%I"` and the script is:

```
#!/bin/bash

wait 5

curl -XPOST cs-tv:2468/api/webhook \
  -H 'Content-Type: application/json' \
  --data '{"infoHash":"'"${1}"'"}'

exit 0
```

I may drop the wait 5 later. I'm trying to avoid the new torrents being set to the 'import pending to Sonarr' category and instead the 'imported already to Sonarr' category.
